cpp/ql/lib/semmle/code/cpp/rangeanalysis/new/internal/semantic/analysis/RangeAnalysisConstantSpecific.qll

```diff
@@ -74,7 +74,10 @@ module CppLangImplConstant implements LangSig<FloatDelta> {
   /**
    * Holds if `e >= bound` (if `upper = false`) or `e <= bound` (if `upper = true`).
    */
-  predicate hasConstantBound(SemExpr e, float bound, boolean upper) { none() }
+  predicate hasConstantBound(SemExpr e, float bound, boolean upper, SemReason reason) {
+    semHasConstantBoundConstantSpecific(e, bound, upper) and
+    reason instanceof SemTypeReason
```

Comment:
I've trouble following this. Why os this restricted to `SemTypeReason` only?

Reply:
Because `semHasConstantBoundConstantSpecific` only infers bounds based on type information. It's probably possible to push this condition into `semHasConstantBoundConstantSpecific`, but that then requires that `semHasConstantBoundConstantSpecific` knows about the `SemReason` type which it currently doesn't since that predicate is defined outside the `CppLangImplConstant` module (which instantiates those types).

Does that make sense?

Reply:
So zero bounds are not constant?

Reply:
`semHasConstantBoundConstantSpecific` just forwards to `hasConstantBoundConstantSpecific` (which is the predicate added [here](https://github.com/github/codeql/pull/13783/commits/050a563580aade23ed1d2e2b755921953c9c38a3#diff-f66e5037244aec24348feede9c60d2584ef91c6b045466f21c6f307d72107008R461)). The only thing that adds is the type bounds inferred from type conversions.

Reply:
We briefly discussed this face-to-face. This is the correct thing to do here because as Mathias writes:

> `semHasConstantBoundConstantSpecific` only infers bounds based on type information

We could extend this later and push down the `reason` argument, but that is not of immediate concern here.